pkg/imgpkg/bundle/images_refs.go

```diff
@@ -0,0 +1,188 @@
+// Copyright 2020 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package bundle
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
+	ctlimg "github.com/k14s/imgpkg/pkg/imgpkg/image"
+	"github.com/k14s/imgpkg/pkg/imgpkg/lockconfig"
+)
+
+//go:generate go run github.com/maxbrunsfeld/counterfeiter/v6 github.com/cppforlife/go-cli-ui/ui.UI
+
+type ImageRef struct {
+	lockconfig.ImageRef
+	IsBundle *bool
+}
+
+func NewImageRef(imgRef lockconfig.ImageRef, isBundle bool) ImageRef {
+	return ImageRef{ImageRef: imgRef, IsBundle: &isBundle}
+}
+
+func (i ImageRef) DeepCopy() ImageRef {
+	return ImageRef{
+		ImageRef: i.ImageRef.DeepCopy(),
+		IsBundle: i.IsBundle,
+	}
+}
+func (i *ImageRef) updateRepo(relativeToRepo string) *ImageRef {
+	i.Image = i.replaceImageRepo(relativeToRepo)
+	return i
+}
+func (i *ImageRef) replaceImageRepo(relativeToRepo string) string {
+	imgParts := strings.Split(i.Image, "@")
+	if len(imgParts) != 2 {
+		panic(fmt.Sprintf("Internal inconsistency: The provided image URL '%s' does not contain a digest", i.Image))
+	}
+	updatedRepo := relativeToRepo + "@" + imgParts[1]
+	return updatedRepo
+}
+
+type ImageRefs struct {
+	refs                 []ImageRef
+	imageLocationsConfig *ImageLocationsConfig
+	originalImagesLock   lockconfig.ImagesLock
+
+	*sync.Mutex
+}
+
+func NewEmptyImageRefs() ImageRefs {
+	return ImageRefs{
+		Mutex: &sync.Mutex{},
+	}
+}
+
+func NewImageRefs(imagesLock lockconfig.ImagesLock, imagesLockLocationConfig ImagesLockLocationsConfig) (ImageRefs, error) {
+	imageRefs := ImageRefs{
+		Mutex:              &sync.Mutex{},
+		originalImagesLock: imagesLock,
+	}
+	for _, lockImgRef := range imagesLock.Images {
+		imageRefs.AddImagesRef(ImageRef{
+			ImageRef: lockImgRef,
+			IsBundle: nil,
+		})
+	}
+
+	err := imageRefs.syncImageRefsWithLocationConfig(imagesLockLocationConfig)
+	if err != nil {
+		return ImageRefs{}, err
+	}
+	return imageRefs, nil
+}
+
+func (i *ImageRefs) LocalizeToBundle(relativeToRepo string) {
+	for _, imgRef := range i.refs {
+		imgRef.AddLocation(imgRef.replaceImageRepo(relativeToRepo))
+		i.AddImagesRef(imgRef)
+	}
+}
+
+func (i *ImageRefs) UpdateRelativeToRepo(imgRetriever ctlimg.ImagesMetadata, relativeToRepo string) (bool, error) {
+	if i.imageLocationsConfig != nil {
+		i.LocalizeToBundle(relativeToRepo)
+		return true, nil
+	}
+
+	for _, ref := range i.refs {
+		image, err := name.ParseReference(ref.updateRepo(relativeToRepo).Image)
```

Comment:
The ref in the for loop is copied by value. `ref.updateRepo` does 'mutate' a locally copied ref value, but `i.refs` does not contain that change.

if we return false because an image is found to not be relocated... then `imagerefs` is not mutated at all.

Reply:
i see. given this explanation whats the reasoning behind having updateRepo at all vs just doing something like `imgRef.replaceImageRepo(relativeToRepo)` like you were doing elsewhere?

Reply:
`updateRepo` calls `imgRef.replaceImageRepo(relativeToRepo)` to update / mutate the .Image on the imgref object.

(as per our below convo) I'm going to move `imgRef.replaceImageRepo` to instead be a package level function that accepts 2 arguments.